src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,683 @@
+}
+
+InputSelectionResult WalletTxBuilder::ResolveInputsAndPayments(
+        const ZTXOSelector& selector,
+        SpendableInputs& spendableMut,
+        const std::vector<Payment>& payments,
+        const CChain& chain,
+        TransactionStrategy strategy,
+        CAmount fee,
+        int anchorHeight) const
+{
+    LOCK2(cs_main, wallet.cs_wallet);
+
+    // Determine the target totals
+    CAmount sendAmount{0};
+    for (const auto& payment : payments) {
+        sendAmount += payment.GetAmount();
+    }
+    CAmount targetAmount = sendAmount + fee;
+
+    // This is a simple greedy algorithm to attempt to preserve requested
+    // transactional privacy while moving as much value to the most recent pool
+    // as possible.  This will also perform opportunistic shielding if the
+    // transaction strategy permits.
+
+    CAmount maxSaplingAvailable = spendableMut.GetSaplingTotal();
+    CAmount maxOrchardAvailable = spendableMut.GetOrchardTotal();
+    uint32_t orchardOutputs{0};
+
+    // we can only select Orchard addresses if there are sufficient non-Sprout
+    // funds to cover the total payments + fee.
+    bool canResolveOrchard =
+        params.GetConsensus().NetworkUpgradeActive(anchorHeight, Consensus::UPGRADE_NU5)
+        && spendableMut.Total() - spendableMut.GetSproutTotal() >= targetAmount;
+    std::vector<ResolvedPayment> resolvedPayments;
+    std::optional<AddressResolutionError> resolutionError;
+    for (const auto& payment : payments) {
+        std::visit(match {
+            [&](const CKeyID& p2pkh) {
+                if (strategy.AllowRevealedRecipients()) {
+                    resolvedPayments.emplace_back(
+                            std::nullopt, p2pkh, payment.GetAmount(), payment.GetMemo(), false);
+                } else {
+                    resolutionError = AddressResolutionError::TransparentRecipientNotAllowed;
+                }
+            },
+            [&](const CScriptID& p2sh) {
+                if (strategy.AllowRevealedRecipients()) {
+                    resolvedPayments.emplace_back(
+                            std::nullopt, p2sh, payment.GetAmount(), payment.GetMemo(), false);
+                } else {
+                    resolutionError = AddressResolutionError::TransparentRecipientNotAllowed;
+                }
+            },
+            [&](const SproutPaymentAddress&) {
+                resolutionError = AddressResolutionError::SproutRecipientsNotSupported;
+            },
+            [&](const SaplingPaymentAddress& addr) {
+                if (strategy.AllowRevealedAmounts() || payment.GetAmount() < maxSaplingAvailable) {
+                    resolvedPayments.emplace_back(
+                            std::nullopt, addr, payment.GetAmount(), payment.GetMemo(), false);
+                    if (!strategy.AllowRevealedAmounts()) {
+                        maxSaplingAvailable -= payment.GetAmount();
+                    }
```

Comment:
This is confusing to me. If both `strategy.AllowRevealedAmounts()` and `payment.GetAmount() <= maxSaplingAvailable`, what do we do?

Reply:
If we `AllowRevealedAmounts`, then we don’t test `maxSaplingAvailable`, so whether or not this decrement happens is irrelevant.

However, in the later PR that restructures how we handle the PrivacyPolicy, we (mostly) don’t check `AllowRevealedAmounts` when resolving payments, so we do this decrement (and the corresponding check) unconditionally in that PR.